### ISSUE TITLE
rlhf训练进度条一开始直接卡死

### ISSUE THREAD

**Issue by rachelsyr** [modelscope/ms-swift#3827]
多卡训练cpo、simpo、orpo进度条一开始会出现卡死现象，gpu利用率变成100%，卡死一段时间后报错：
[PG ID 0 PG GUID 0(default_pg) Rank 1] Timeout at NCCL work: 11, last enqueued NCCL work: 11, last completed NCCL work: 10.

[Rank 1] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data

Process group watchdog thread terminated with exception: [Rank 1] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=11, OpType=_ALLGATHER_BASE, NumelIn=1, NumelOut=2, Timeout(ms)=1800000)
报错位置在oss, metrics = self.get_batch_loss_metrics(model, inputs, train_eval="train")

运行脚本
CUDA_VISIBLE_DEVICES=0,1 \
NPROC_PER_NODE=2 \
swift rlhf \
    --rlhf_type simpo \
    --model "/data/home/models/Qwen2.5-1.5B-Instruct" \
    --train_type full \
    --dataset "/data/home/datasets/Human-Like-DPO-Dataset#5000" \
    --torch_dtype bfloat16 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 1 \
    --learning_rate 1e-5 \
    --gradient_accumulation_steps 4 \
    --eval_steps 100 \
    --save_steps 100 \
    --save_total_limit 5 \
    --logging_steps 5 \
    --max_length 1024 \
    --output_dir output \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --deepspeed zero2 \
运行环境：cuda12.8 A100
absl-py                           2.1.0
accelerate                        1.6.0
addict                            2.4.0
aenum                             3.1.15
aiofiles                          23.2.1
aiohappyeyeballs                  2.6.1
aiohttp                           3.11.14
aiohttp-cors                      0.8.0
aiosignal                         1.3.2
airportsdata                      20250224
aliyun-python-sdk-core            2.16.0
aliyun-python-sdk-kms             2.16.5
altair                            5.5.0
annotated-types                   0.7.0
anthropic                         0.49.0
antlr4-python3-runtime            4.13.2
anyio                             4.9.0
argon2-cffi                       23.1.0
argon2-cffi-bindings              21.2.0
arrow                             1.3.0
arxiv                             2.2.0
astor                             0.8.1
asttokens                         3.0.0
async-lru                         2.0.5
attrdict                          2.0.1
attrs                             25.3.0
babel                             2.17.0
beautifulsoup4                    4.13.3
binpacking                        1.5.2
bitsandbytes                      0.45.3
blake3                            1.0.4
bleach                            6.2.0
blinker                           1.9.0
blis                              0.7.11
boto3                             1.37.30
botocore                          1.37.30
Brotli                            1.1.0
cachetools                        5.5.2
catalogue                         2.0.10
certifi                           2025.1.31
cffi                              1.17.1
chardet                           5.2.0
charset-normalizer                3.4.1
click                             8.1.8
cloudpathlib                      0.16.0
cloudpickle                       3.1.1
colorama                          0.4.6
colorful                          0.5.6
colorlog                          6.9.0
comm                              0.2.2
compressed-tensors                0.9.1
confection                        0.1.5
contourpy                         1.3.1
cpm-kernels                       1.0.11
crcmod                            1.7
cryptography                      44.0.2
cupy-cuda12x                      13.4.1
cycler                            0.12.1
cymem                             2.0.11
dacite                            1.9.2
DataProperty                      1.1.0
datasets                          3.2.0
debugpy                           1.8.13
decorator                         5.2.1
decord                            0.6.0
deepspeed                         0.15.4
defusedxml                        0.7.1
depyf                             0.18.0
dill                              0.3.8
diskcache                         5.6.3
distlib                           0.3.9
distro                            1.9.0
dnspython                         2.7.0
docker-pycreds                    0.4.0
duckduckgo_search                 5.3.1b1
einops                            0.8.1
email_validator                   2.2.0
et_xmlfile                        2.0.0
evalscope                         0.13.2
evaluate                          0.4.3
executing                         2.2.0
fastapi                           0.115.11
fastapi-cli                       0.0.7
fastjsonschema                    2.21.1
fastrlock                         0.8.3
feedparser                        6.0.11
ffmpy                             0.5.0
filelock                          3.18.0
fire                              0.7.0
flake8                            7.1.2
flash_attn                        2.7.4.post1
fonttools                         4.56.0
fqdn                              1.5.1
frozenlist                        1.5.0
fsspec                            2024.9.0
func_timeout                      4.3.5
future                            1.0.0
fuzzywuzzy                        0.18.0
gguf                              0.10.0
gitdb                             4.0.12
GitPython                         3.1.44
google-api-core                   2.24.2
google-auth                       2.38.0
googleapis-common-protos          1.69.2
gradio                            5.22.0
gradio_client                     1.8.0
griffe                            0.49.0
groovy                            0.1.2
grpcio                            1.71.0
h11                               0.14.0
h2                                4.2.0
h5py                              3.13.0
hf_transfer                       0.1.9
hjson                             3.1.0
hpack                             4.1.0
httpcore                          1.0.7
httptools                         0.6.4
httpx                             0.28.1
huggingface-hub                   0.29.3
human-eval                        1.0.3
hyperframe                        6.1.0
idna                              3.10
imageio                           2.37.0
immutabledict                     4.2.1
importlib_metadata                8.6.1
iniconfig                         2.0.0
inquirerpy                        0.3.4
interegular                       0.3.3
ipykernel                         6.29.5
ipython                           9.1.0
ipython_pygments_lexers           1.1.1
ipywidgets                        8.1.5
isoduration                       20.11.0
isort                             6.0.1
jedi                              0.19.2
jieba                             0.42.1
Jinja2                            3.1.6
jiter                             0.9.0
jmespath                          0.10.0
joblib                            1.4.2
json5                             0.12.0
jsonlines                         4.0.0
jsonpointer                       3.0.0
jsonschema                        4.23.0
jsonschema-specifications         2024.10.1
jupyter                           1.1.1
jupyter_client                    8.6.3
jupyter-console                   6.6.3
jupyter_core                      5.7.2
jupyter-events                    0.12.0
jupyter-lsp                       2.2.5
jupyter_server                    2.15.0
jupyter_server_terminals          0.5.3
jupyterlab                        4.3.6
jupyterlab_pygments               0.3.0
jupyterlab_server                 2.27.3
jupyterlab_widgets                3.0.13
kiwisolver                        1.4.8
lagent                            0.2.4
langcodes                         3.5.0
langdetect                        1.0.9
language_data                     1.3.0
lark                              1.2.2
latex2mathml                      3.77.0
latex2sympy2                      1.9.1
latex2sympy2_extended             1.10.1
Levenshtein                       0.27.1
liger_kernel                      0.5.3
lighteval                         0.6.0.dev0     /data/home/workspace/lighteval
linkify-it-py                     2.0.3
llvmlite                          0.43.0
lm-format-enforcer                0.10.11
lmdeploy                          0.7.2.post1
lxml                              5.3.1
marisa-trie                       1.2.1
Markdown                          3.7
markdown-it-py                    3.0.0
markdown2                         2.5.3
MarkupSafe                        3.0.2
math-verify                       0.5.2
matplotlib                        3.10.1
matplotlib-inline                 0.1.7
mbstrdecoder                      1.1.4
mccabe                            0.7.0
mdit-py-plugins                   0.4.2
mdurl                             0.1.2
memray                            1.16.0
mistral_common                    1.5.4
mistune                           3.1.3
mmengine                          0.10.7
mmengine-lite                     0.10.7
modelscope                        1.24.0
more-itertools                    10.6.0
mpmath                            1.3.0
ms-opencompass                    0.1.6
ms_swift                          3.3.0.dev0     /data/home/workspace/ms-swift
ms-vlmeval                        0.0.16
msgpack                           1.1.0
msgspec                           0.19.0
multidict                         6.2.0
multiprocess                      0.70.16
murmurhash                        1.0.12
narwhals                          1.34.1
nbclient                          0.10.2
nbconvert                         7.16.6
nbformat                          5.10.4
nest-asyncio                      1.6.0
networkx                          3.4.2
nh3                               0.2.21
ninja                             1.11.1.3
nltk                              3.9.1
notebook                          7.3.3
notebook_shim                     0.2.4
numba                             0.60.0
numpy                             1.26.4
nvidia-cublas-cu12                12.4.5.8
nvidia-cuda-cupti-cu12            12.4.127
nvidia-cuda-nvrtc-cu12            12.4.127
nvidia-cuda-runtime-cu12          12.4.127
nvidia-cudnn-cu12                 9.1.0.70
nvidia-cufft-cu12                 11.2.1.3
nvidia-curand-cu12                10.3.5.147
nvidia-cusolver-cu12              11.6.1.9
nvidia-cusparse-cu12              12.3.1.170
nvidia-ml-py                      12.570.86
nvidia-nccl-cu12                  2.21.5
nvidia-nvjitlink-cu12             12.4.127
nvidia-nvtx-cu12                  12.4.127
omegaconf                         2.0.0
openai                            1.66.3
OpenCC                            1.1.9
opencensus                        0.11.4
opencensus-context                0.1.3
opencv-python                     4.11.0.86
opencv-python-headless            4.11.0.86
openpyxl                          3.1.5
orjson                            3.10.15
oss2                              2.19.1
outlines                          0.1.11
outlines_core                     0.1.26
overrides                         7.7.0
packaging                         24.2
pandas                            2.2.3
pandocfilters                     1.5.1
parameterized                     0.9.0
parso                             0.8.4
partial-json-parser               0.2.1.1.post5
pathvalidate                      3.2.3
peft                              0.14.0
pexpect                           4.9.0
pfzy                              0.3.4
phx-class-registry                4.1.0
pillow                            11.1.0
pip                               24.3.1
platformdirs                      4.3.6
pluggy                            1.5.0
portalocker                       3.1.1
preshed                           3.0.9
prettytable                       3.16.0
prometheus_client                 0.21.1
prometheus-fastapi-instrumentator 7.0.2
prompt_toolkit                    3.0.50
propcache                         0.3.0
proto-plus                        1.26.1
protobuf                          3.20.3
psutil                            7.0.0
ptyprocess                        0.7.0
pure_eval                         0.2.3
py-cpuinfo                        9.0.0
py-spy                            0.4.0
pyarrow                           19.0.1
pyasn1                            0.6.1
pyasn1_modules                    0.4.1
pybind11                          2.13.6
pycodestyle                       2.12.1
pycountry                         24.6.1
pycparser                         2.22
pycryptodome                      3.22.0
pydantic                          2.10.6
pydantic_core                     2.27.2
pydantic-settings                 2.8.1
pydeck                            0.9.1
pydub                             0.25.1
pyflakes                          3.2.0
Pygments                          2.19.1
pynvml                            12.0.0
pyparsing                         3.2.1
pypinyin                          0.54.0
pysqlite3                         0.5.4
pytablewriter                     1.2.1
pytest                            8.3.5
python-dateutil                   2.9.0.post0
python-dotenv                     1.0.1
python-json-logger                3.3.0
python-Levenshtein                0.27.1
python-multipart                  0.0.20
pytz                              2025.1
PyYAML                            6.0.2
pyzmq                             26.3.0
rank-bm25                         0.2.2
RapidFuzz                         3.13.0
ray                               2.40.0
referencing                       0.36.2
regex                             2024.11.6
requests                          2.32.3
rfc3339-validator                 0.1.4
rfc3986-validator                 0.1.1
rich                              13.9.4
rich-toolkit                      0.13.2
rouge                             1.0.1
rouge-chinese                     1.0.3
rouge_score                       0.1.2
rpds-py                           0.23.1
rsa                               4.9
ruff                              0.11.0
s3transfer                        0.11.4
sacrebleu                         2.5.1
safehttpx                         0.1.6
safetensors                       0.5.3
scikit-learn                      1.6.1
scipy                             1.15.2
seaborn                           0.13.2
semantic-version                  2.10.0
Send2Trash                        1.8.3
sentence-transformers             4.0.2
sentencepiece                     0.2.0
sentry-sdk                        2.23.1
setproctitle                      1.3.5
setuptools                        69.5.1
sgmllib3k                         1.0.0
shellingham                       1.5.4
shortuuid                         1.0.13
simplejson                        3.20.1
six                               1.17.0
smart-open                        6.4.0
smmap                             5.0.2
sniffio                           1.3.1
socksio                           1.0.0
sortedcontainers                  2.4.0
soupsieve                         2.6
spacy                             3.7.2
spacy-legacy                      3.0.12
spacy-loggers                     1.0.5
srsly                             2.5.1
stack-data                        0.6.3
starlette                         0.46.1
streamlit                         1.44.1
sty                               1.0.6
svgwrite                          1.4.3
swankit                           0.1.7
swanlab                           0.5.5
sympy                             1.13.1
tabledata                         1.3.4
tabulate                          0.9.0
tcolorpy                          0.1.7
tenacity                          9.1.2
tensorboard                       2.19.0
tensorboard-data-server           0.7.2
termcolor                         2.3.0
terminado                         0.18.1
textual                           2.1.2
thinc                             8.2.5
threadpoolctl                     3.6.0
tiktoken                          0.9.0
timeout-decorator                 0.5.0
tinycss2                          1.4.0
tokenizers                        0.21.1
toml                              0.10.2
tomlkit                           0.13.2
torch                             2.5.1
torchaudio                        2.5.1
torchvision                       0.20.1
tornado                           6.4.2
tqdm                              4.67.1
traitlets                         5.14.3
transformers                      4.49.0
transformers-stream-generator     0.0.5
triton                            3.1.0
trl                               0.15.2
typepy                            1.3.4
typer                             0.15.2
types-python-dateutil             2.9.0.20241206
typing_extensions                 4.12.2
tzdata                            2025.1
uc-micro-py                       1.0.3
uri-template                      1.3.0
urllib3                           2.3.0
uvicorn                           0.34.0
uvloop                            0.21.0
validators                        0.34.0
virtualenv                        20.29.3
vllm                              0.7.3
wandb                             0.19.8
wasabi                            1.1.3
watchdog                          6.0.0
watchfiles                        1.0.4
wavedrom                          2.0.3.post3
wcwidth                           0.2.13
weasel                            0.3.4
webcolors                         24.11.1
webencodings                      0.5.1
websocket-client                  1.8.0
websockets                        15.0.1
Werkzeug                          3.1.3
widgetsnbextension                4.0.13
word2number                       1.1
wrapt                             1.17.2
xformers                          0.0.28.post3
xgrammar                          0.1.11
XlsxWriter                        3.2.2
xtuner                            0.1.11
xxhash                            3.5.0
yapf                              0.43.0
yarl                              1.18.3
zipp                              3.21.0
zstandard                         0.23.0

上述算法卡死问题可以稳定复现，而类似的偏好算法dpo rm，以及sft grpo没有这个问题；之前我用同样的数据集采样5000后跑过cpo、simpo、orpo也没问题，换了个数据集后就卡死了，恢复到原来成功运行的数据集后依然卡死

**Comment by hjh0119** [comment]
I think that https://github.com/huggingface/trl/pull/3069 should have fixed this issue. Could you please update trl and try again?

**Comment by rachelsyr** [comment]
> I think that [huggingface/trl#3069](https://github.com/huggingface/trl/pull/3069) should have fixed this issue. Could you please update trl and try again?

Thank you very much. I resolved this problem by building trl from source.
